Feature discard upstream HLS

stevemayhew opened the pull request:
Pull request is not ready to merge.  The HLS libraries `HlsSampleStreamWrapper` needs a way to manage the queued samples matching up with the `HlsMediaChunk` that will be discarded.  See notes in the diffs below.

stevemayhew committed:
add support SequenceableLoader.reevaluateBuffer() for HLS

DASH implements this feature, extend the feature for HLS as well.  First change just drops video samples.
For demuxed audio the audio samples will continue to play out to match the
dropped video, so need to keep indexes in all the sample queues related to a chunk and discard them all.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java:
```diff
@@ -433,6 +433,13 @@ public boolean onPlaylistError(Uri playlistUrl, long blacklistDurationMs) {
+    }
+    return trackSelection.evaluateQueueSize(playbackPositionUs, queue);
+  }
+
```
I'm looking at actually implementing `ChunkSource` interface.  This looks mostly possible for HLS library except that the `ChunkHolder` is a final class....  Making it an interface and allowing `HlsChunkSource` to implement it looks like the best path.

One of the reasons I'm putting this pull up early is to get feed back on the history and desired architecture direction

tonihei replied:
HLS does not implement some of the common interfaces like ChunkSource and ChunkSampleStream, for historical reasons and also because HLS has a slightly more complicated media structure that is different from DASH and Smooth Streaming. It would be easier to keep it that way for now even if it means to copy paste code. We can consolidate that later.

stevemayhew replied:
Got it.  Yes the code is similar.  Main difference is how HLS manages lists of `SampleStreams`.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -652,7 +652,15 @@ public boolean continueLoading(long positionUs) {
+    int preferredQueueSize = chunkSource.getPreferredQueueSize(positionUs, readOnlyMediaChunks);
+    if (currentQueueSize <= preferredQueueSize) {
+      return;
+    }
```
Just the first checkin, see next changes to the class for the balance of the code.  It seems extending `ChunkSampleStream` is a better path for `HlsSampleStreamWrapper`.  Again, if that is the desired architecture I'm fine with taking that path.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -652,7 +652,42 @@ public boolean continueLoading(long positionUs) {
+    // TODO - need a way to discard just the samples owned by the discarded chunks
+    sampleQueues[primarySampleQueueIndex].discardToRead();
+    Util.removeRange(mediaChunks, 0, chunksToDiscard  - 1);
+    return lastDiscardEndTimeUs;
```
This is just the first part of the implementation, because it is not yet possible to extend `ChunkSampleStream` this results in code copy from it.

It seems extending `ChunkSampleStream` is a better path for `HlsSampleStreamWrapper`.  Again, if that is the desired architecture I'm fine with taking that path.

Note the HlsSampleSteamWrapper's `mediaChunks` need to be discarded from index 0 to the index of the first 'good' format sample.  This seems not to be the way of `ChunkSampleStream#discardUpstreamMediaChunksFromIndex()`

Here's a snapshot of the logging output from above from an actual live stream that has up leveled.

````
08-19 16:27:06.749 21933 21967 D HlsSampleStreamWrapper: discardChunksToIndex() -  chunksToDiscard 1 count 4
08-19 16:27:06.750 21933 21967 D HlsSampleStreamWrapper: chunk 5 start/end: 1792000000/1796000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.751 21933 21967 D HlsSampleStreamWrapper: chunk 7 start/end: 1796000000/1800000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.751 21933 21967 D HlsSampleStreamWrapper: chunk 8 start/end: 1800000000/1804000000 format: Format(2, null, application/x-mpegURL, null, mp4a.40.2,avc1.64001F, 2100000, null, [960, 540, -1.0], [-1, -1])
08-19 16:27:06.752 21933 21967 D HlsSampleStreamWrapper: chunk 10 start/end: 1800000000/1804000000 format: Format(4, null, application/x-mpegURL, null, mp4a.40.2,avc1.640020, 8250000, null, [1280, 720, -1.0], [-1, -1])
````

So the samples and chunks with UID's 5,7,8 need to be discarded.

tonihei replied:
To make that work, HlsMediaChunk needs to store a list of the first sample index for each track (similar to BaseMediaChunk). But I haven't checked in detail if and how we can do this.

stevemayhew replied:
This has been done... See the latest checkin

stevemayhew commented:
This is the feature request for this work: https://github.com/google/ExoPlayer/issues/6322

stevemayhew committed:
Update to save all sample queue write indexes

Finailzed logic to update the `SampleQueue` write positions (first index) to push these into `HlsMediaChunk` when the track is initially created (from the Extractor output) and as each new chunk is queued to load (`init()` callback).

Add lots of debuging prints that can come out for the final merge.

Code is very close to a clone of `ChunkSampleStream`.

stevemayhew commented:
I am testing this now.  It is mostly working, but I see some glitches.   The test case I have is a demux ed stream (so there are two `HlsSampleStreamWrapper`, one for the demux single audio track and the other for the video.  The stream is live and as you can see in `PlayerActivity` I've set the retain time very aggressively (2seconds).

The code is ready to review, but I'm still seeing issues in my testing so don't merge it in yet.

stevemayhew commented on demos/main/src/main/java/com/google/android/exoplayer2/demo/PlayerActivity.java:
```diff
@@ -352,7 +357,12 @@ private void initializePlayer() {
+            DEFAULT_MIN_DURATION_FOR_QUALITY_INCREASE_MS,
+            DEFAULT_MAX_DURATION_FOR_QUALITY_DECREASE_MS,
+            2000,
+            DEFAULT_BANDWIDTH_FRACTION
```
2 seconds pretty much forces this to happen on initial startup of live stream.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -211,6 +213,12 @@ public static HlsMediaChunk createInstance(
+   * media chunk
+   */
+  private int [] firstSampleIndexes = new int[0];
+
```
Keep all the SampleStreams as discovered in the MPEG-TS parse.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -280,6 +288,20 @@ public boolean isLoadCompleted() {
   }
 
+  /**
+   * Return the index of the first sample from the primary sample stream for this media chunk
```
samples, I'll fix this comment

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -652,9 +653,93 @@ public boolean continueLoading(long positionUs) {
+
+    dumpCurrentChunkList();
+
+    return firstRemovedChunk;
```
This code is near a literal copy now of `ChunkSampleStream`.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -652,9 +653,93 @@ public boolean continueLoading(long positionUs) {
+          + " start/end: " + chunk.startTimeUs + "/" + chunk.endTimeUs + " primary sample index: "
+          + chunk.getFirstSampleIndexes()[primarySampleQueueIndex] + " format: "+chunk.trackFormat);
+    }
+  }
```
Just for my debugging.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -652,9 +653,93 @@ public boolean continueLoading(long positionUs) {
+    }
+    return haveRead;
+  }
   // Loader.Callback implementation.
```
Again nearly identical to `ChunkSampleStream`

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -849,6 +937,8 @@ public TrackOutput track(int id, int type) {
     sampleQueues[trackCount] = trackOutput;
+    HlsMediaChunk mediaChunk = findChunkMatching(chunkUid);
+    mediaChunk.setFirstSampleIndex(trackCount, 0);
     sampleQueueReaders = Arrays.copyOf(sampleQueueReaders, trackCount + 1);
```
When each elementary stream is discovered and the SampleQueue is created, this is needed for the first chunks loaded right after the initial Extractor is created.

stevemayhew commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -898,8 +988,36 @@ public void setSampleOffsetUs(long sampleOffsetUs) {
+
+    return writePositions;
+  }
+
```
This method and it's call are no longer needed.  Will remove

stevemayhew replied:
removed

stevemayhew commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/DebugTextViewHelper.java:
```diff
@@ -18,25 +18,46 @@
 import android.annotation.SuppressLint;
```
I can leave this file out of the review.  Changes are to expose the upstream as well as the downstream level and the current playback position in US.

stevemayhew committed:
Remove unused methods, comment cleanup.

stevemayhew commented:
I'm currently testing by forcing poor choices on the initial bitrate for a live stream:

```java

      DefaultBandwidthMeter bandwidthMeter = new DefaultBandwidthMeter.Builder(getApplicationContext())
          .setSlidingWindowMaxWeight(DefaultBandwidthMeter.DEFAULT_SLIDING_WINDOW_MAX_WEIGHT).build();
      bandwidthMeter.setNetworkTypeOverride(C.NETWORK_TYPE_3G);
```

I'm also adjusting the sliding weight to cause it to adapt more slowly to the new bandwidth discovered.

I can provide URLs of live broadcasts here that I'm testing with if you give me an email I can send them to you.

stevemayhew commented:
Sorry, I can get back to you tomorrow.  If the tracks have same GOP rate
(required for HLS) this should work fine. Perhaps we need to limit it based
on the independent playlist supported flag?



On Mon, Sep 30, 2019 at 6:04 PM Google Open Source Bot <
notifications@github.com> wrote:

> Hey @stevemayhew <https://github.com/stevemayhew>. We need more
> information to resolve this issue but there hasn't been an update in 14
> days. I'm marking the issue as stale and if there are no new updates in the
> next 7 days I will close it automatically.
>
> If you have more information that will help us get to the bottom of this,
> just add a comment!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/6321?email_source=notifications&email_token=AADBF6AGLTL5QNKDLPHURPDQMKO3PA5CNFSM4INJRJD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD77R7ZY#issuecomment-536813543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADBF6GOEWVWSWP6KD3RFTTQMKO3PANCNFSM4INJRJDQ>
> .
>

tonihei commented:
Sorry for the delay, but just checking - is this ready for review now or still being worked on?

stevemayhew commented:
We are testing this code in our own branches.  I’ll rebase it to keep up to
date with dev-v2.  This does make channel changes faster for us

On Mon, Nov 25, 2019 at 6:05 PM Google Open Source Bot <
notifications@github.com> wrote:

> Hey @stevemayhew <https://github.com/stevemayhew>. We need more
> information to resolve this issue but there hasn't been an update in 14
> days. I'm marking the issue as stale and if there are no new updates in the
> next 7 days I will close it automatically.
>
> If you have more information that will help us get to the bottom of this,
> just add a comment!
>
> —
>
>
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/6321?email_source=notifications&email_token=AADBF6HLVAUPZ6ZKLMZXT2DQVR777A5CNFSM4INJRJD2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFEOJOI#issuecomment-558425273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBF6CGAFB4AFDZXHPZ3G3QVR777ANCNFSM4INJRJDQ>
> .
>

stevemayhew commented:
Rebased this to the latest `dev-v2` and re-tested, drop me a note if you want a private live URL you can test this against.

Simply forcing a low initial bandwidth estimate and discarding aggressively on level shift takes the time to full bit rate on initial channel change down by 2 seconds (the size of the discard).  I can send a patch, but basically change `PlayerActivity` track selection and bandwidth meter as:

```java
      if (abrAlgorithm == null || ABR_ALGORITHM_DEFAULT.equals(abrAlgorithm)) {
        trackSelectionFactory = new AdaptiveTrackSelection.Factory(
            DEFAULT_MIN_DURATION_FOR_QUALITY_INCREASE_MS,
            DEFAULT_MAX_DURATION_FOR_QUALITY_DECREASE_MS,
            2000,
            DEFAULT_BANDWIDTH_FRACTION
        );
      }
```
...
```java
      player =
          new SimpleExoPlayer.Builder(/* context= */ this, renderersFactory)
              .setTrackSelector(trackSelector)
              .setBandwidthMeter(new DefaultBandwidthMeter.Builder(this)
                  .setInitialBitrateEstimate(500)
                  .build())
              .build();
      player.addListener(new PlayerEventListener());
```
.

biespb commented:
[https://en.av01.tv/](url)

biespb reviewed (commented):
[snip]

stevemayhew commented:
I'm not rushing to get this merged in...  We are finding architectural issues (`ConcurrentModificationException`) with this change that I am still looking at the best way to fix.  I would love some  guidance from the ExoPlayer architecture team (@tonihei @ojw28 )

The issue is both the `HlsMediaChunk` list and the `SampleQueue` array are modified by different threads.  This is true before this change as well.

For example, the player thread modifies the list here when reading samples:
https://github.com/google/ExoPlayer/blob/826083db923f26f2000ca42f7b54b9531726d713/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java#L543

And the loading thread of course modifies it here (one of many examples):  https://github.com/google/ExoPlayer/blob/826083db923f26f2000ca42f7b54b9531726d713/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java#L774

This comment:  https://github.com/google/ExoPlayer/blob/826083db923f26f2000ca42f7b54b9531726d713/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java#L818

Is no longer true based on the code (the `Loadable` calls the `init()` method from the `Loader`'s run loop here:  https://github.com/google/ExoPlayer/blob/826083db923f26f2000ca42f7b54b9531726d713/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java#L309 and it is iterating an array that seems more 'owned' by the `HlsSampleStreamWrapper` itself.

The synchronizing by avoidance (make sure the loading threads are not running before calling any method that accesses structures in the wrapper) seems to be breaking down, and I'm not sure which method to fix it best matches the direction the overall architecture for Chunk/Sample loading interfacing with the player thread is headed.

The options I see are:
1. Wrap the lists and arrays that are modified/accessed by both threads with `Collections.synchronizedList()` - big hammer, not obvious what is going on
2. Extract methods for the sections that modify and synchronize them with the player thread (using the `HlsSampleStreamWrapper`
3. Lift an interface for methods in the `HlsSampleStreamWrapper` called by the loader thread (https://github.com/google/ExoPlayer/blob/826083db923f26f2000ca42f7b54b9531726d713/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java#L217) and synchronize them and any other methods that access fields shared by these methods with `HlsSampleStreamWrapper` object (the player thread effectively)

AquilesCanta commented:
> The issue is both the HlsMediaChunk list and the SampleQueue array are modified by different threads. This is true before this change as well.

I think both modifications quoted above happen on the player thread.

If you think there are threading concerns we should address in the existing code, please file a new issue including all the information required in the issue template. We should ensure the existing code is correct in order to make the review of this pull request constrained to the new feature, and not also fixing a problem in the original code. 

Once we agree on the existing code being thread-safe, we can edit the comment above to limit it to the changes we'd like to introduce.

EDIT: Note that there is indeed an issue (already sent a fix for internal review) with the thread access comments around `HlsSampleStreamWrapper.init()`. I don't think they are related to the thread issues you mentioned above, but still worth mentioning.

AquilesCanta commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -222,6 +224,12 @@ public static HlsMediaChunk createInstance(
+   * Index of first sample written for each SampleQueue created from the extraction of this
+   * media chunk
+   */
+  private int [] firstSampleIndexes = new int[0];
```
Super-nit: Here and elsewhere, use don't introduce a space between int and [], so this should be:

`private int[] firstSampleIndexes = new int[0];`

stevemayhew replied:
thanks.. Will do.

AquilesCanta commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsMediaChunk.java:
```diff
@@ -291,6 +299,27 @@ public boolean isLoadCompleted() {
+   * @param firstSampleIndex - index value to store
+   */
+  void setFirstSampleIndex(int streamIndex, int firstSampleIndex) {
+    firstSampleIndexes = Arrays.copyOf(firstSampleIndexes, streamIndex + 1);
```
What do you think should happen if a segment does not include samples for a given track? A valid option would be using C.INDEX_UNSET, or zero, in which case we should adjust the javadocs.

stevemayhew replied:
> EDIT: Note that there is indeed an issue (already sent a fix for internal review) with the thread access comments around `HlsSampleStreamWrapper.init()`. I don't think they are related to the thread issues you mentioned above, but still worth mentioning.

What's the issue number?  Because the method is definitely called from the loader thread, this happens concurrently with sample reading on the player thread so concurrent modifications can happen.

Note, of course you know even iterating an collection in one thread that is being modified in another thread will result in an ConcurrentModificationException.   I'll help address this in the issue discussion if I can.  Just let me know how best I can help.

AquilesCanta replied:
The exception you mentioned shouldn't be related to the init issue. Please file a fresh issue for that, it will be easier for us to debug it.

AquilesCanta commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -696,9 +697,93 @@ public boolean isLoading() {
+   * @param chunkIndex The index of the first chunk to discard.
+   * @return The chunk at given index.
+   */
+  private HlsMediaChunk discardUpstreamMediaChunksFromIndex(int chunkIndex) {
```
We should add a comment in the javadoc saying "Must only be called when there's no active loading thread, or similar, which I think is already fulfilled by reevaluate buffer.

AquilesCanta commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -696,9 +697,93 @@ public boolean isLoading() {
+
+
+    int [] firstSamples = firstRemovedChunk.getFirstSampleIndexes();
+    for (int i=0; i < sampleQueues.length; i++) {
```
Super-nit: This should be 

`int i = 0; ...`

Ditto elsewhere.

AquilesCanta commented on library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:
```diff
@@ -696,9 +697,93 @@ public boolean isLoading() {
+  }
+
+
+  /** Returns whether samples have been read from primary sample queue of the indicated chunk */
```
Could you clarify why the primary sample queue is enough, as opposed to any sample queue?

ojw28 commented:
@stevemayhew @AquilesCanta - Just wondering what the current status of this PR is? We have another external contributor who has some changes that build on top of this, so it would be good to figure out what we need to do here and/or what the open issues are. Thanks!

stevemayhew commented:
Oliver (@ojw28) I need to rebase and test again.  With short segments
(I-Frame for example) I saw `ConcurrentModificationException` on the list
of `MediaChunk`.

You saw my earlier notes as to why this is possible.  My feeling is the
loader threads should never touch this list, the changes to `init()`
recently might have fixed it so they don’t, I’ll have to look at it again


On Tue, Apr 14, 2020 at 9:21 AM Oliver Woodman <notifications@github.com>
wrote:

> @stevemayhew <https://github.com/stevemayhew> @AquilesCanta
> <https://github.com/AquilesCanta> - Just wondering what the current
> status of this PR is? We have another external contributor who has some
> changes that build on top of this, so it would be good to figure out what
> we need to do here and/or what the open issues are. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/ExoPlayer/pull/6321#issuecomment-613540689>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBF6D62E3WCQNG72DZCZDRMSERJANCNFSM4INJRJDQ>
> .
>

stevemayhew commented:
Closing this pull request, it is a duplicate of  pull #7244 which is further along.
